File: cpp/problem_4.cpp
#include <iostream>
#include <string>

bool is_palindrome(const std::string &s) {
    std::string::const_iterator lo = s.begin(), hi = --s.end();
    while (*lo == *hi && lo < hi ) 
    	++lo, --hi;
    return (lo >= hi);
}

int problem_4(const int &lo, const int &hi) {
	for(int i = hi; i >= lo; --i) {
		for(int j = i; j >= lo; --j) {
			int x = i * j;
			if (is_palindrome(std::to_string(x)))
				return x;
		}
	}
}

int main() {
	std::cout << problem_4(100, 999) << std::endl;
	return 0;
}

File: Scala/problem4.scala
def isPalindrome(str: String): Boolean = {
  (str == str.reverse)
}

def xDigitRange(digitCount: Int): Range = {
  def powerOfTen(n: Int): Int = scala.math.pow(10, n).toInt
	val lower: Int = powerOfTen(digitCount - 1)
	val upper: Int = powerOfTen(digitCount)
	(lower until upper)
}

def problem4(digitCount: Int): Int = {
  val rng = xDigitRange(digitCount)
  lazy val allPalindromes = for {
    i <- rng.reverse
    j <- (i to rng.head by -1)
    prod = i * j
    if isPalindrome(prod.toString)
  } yield prod
  allPalindromes.head
}
